srcts/src/components/sidebar.ts

```diff
@@ -4,20 +4,173 @@ import {
+
+    const childLayouts = container.getElementsByClassName(
+      BslibSidebar.LAYOUT_CLASS
+    );
+
+    if (childLayouts.length > 0) {
+      BslibSidebar._initAutoCollapse(container);
+      return;
+    }
+
+    // This layout is the innermost layout, so we add CSS variables to it, and
+    // any direct ancestor sidebar layouts, that count the number of parent
+    // layouts in this subtree. This is used to ensure the collapse toggles
+    // don't overlap.
+
+    function nextSidebarParent(el: HTMLElement | null): HTMLElement | null {
+      el = el ? el.parentElement : null;
+      if (el && el.classList.contains("main")) {
+        // .bslib-sidebar-layout > .main > .bslib-sidedbar-layout
+        el = el.parentElement;
+      }
+      if (el && el.classList.contains(BslibSidebar.LAYOUT_CLASS)) {
+        return el;
+      }
+      return null;
+    }
+
+    const layouts = [container];
+    let parent = nextSidebarParent(el);
+
+    while (parent) {
+      // Add parent to front of layouts array, so we sort outer -> inner
+      layouts.unshift(parent);
+      parent = nextSidebarParent(el);
+    }
+
+    const count = { left: 0, right: 0 };
+    layouts.forEach(function (x: HTMLElement, i: number): void {
+      x.style.setProperty("--bslib-sidebar-counter", i.toString());
+      const isRight = x.classList.contains("sidebar-right");
+      const thisCount = isRight ? count.right++ : count.left++;
+      x.style.setProperty(
+        "--bslib-sidebar-overlap-counter",
+        thisCount.toString()
+      );
+    });
+
+    BslibSidebar._initAutoCollapse(container);
+  }
+
+  private static _initAutoCollapse(container: HTMLElement): void {
+    if (!container.dataset.sidebarInitAutoCollapse) {
+      return;
+    }
+
+    // If sidebar is marked open='desktop', then close sidebar on mobile
+    const initCollapsed = window
+      .getComputedStyle(container)
+      .getPropertyValue("--bslib-sidebar-js-init-collapsed");
+    if (initCollapsed === "true") {
+      BslibSidebar.toggleCollapse(container, "close");
+    }
+  }
+
+  private static _findLayoutContainer(el: HTMLElement): HTMLElement {
+    if (el.classList.contains(BslibSidebar.LAYOUT_CLASS)) {
+      return el;
+    }
+    const container = el.closest(`.${BslibSidebar.LAYOUT_CLASS}`);
+    if (!container) {
+      throw new Error(
+        `Expected container or direct ancestor with class ${BslibSidebar.LAYOUT_CLASS}`
+      );
+    }
+    return container as HTMLElement;
+  }
+
+  public static sidebarComponents(el: HTMLElement): SidebarComponents {
+    el = BslibSidebar._findLayoutContainer(el);
+
+    // sidebar components
+    const main = el.querySelector(":scope > .main") as HTMLElement;
+    const sidebar = el.querySelector(":scope > .sidebar") as HTMLElement;
+    const toggle = el.querySelector(":scope > .collapse-toggle") as HTMLElement;
+
+    // sidebar state
+    const isClosed = el.classList.contains(BslibSidebar.COLLAPSE_CLASS);
```

Comment:
Maybe it'd be better if this was a function?

```suggestion
    const isClosed = () => el.classList.contains(BslibSidebar.COLLAPSE_CLASS);
```

Reply:
For now I don't see a strong need for this in our current code. But if we switched at some point to using the `new Sidebar(el)` pattern it'd make more sense for this to be a function.